server/src/main/java/org/elasticsearch/search/aggregations/bucket/sampler/random/RandomSamplerAggregationBuilder.java

```diff
@@ -0,0 +1,181 @@
+    }
+
+    @Override
+    protected AggregatorFactory doBuild(
+        AggregationContext context,
+        AggregatorFactory parent,
+        AggregatorFactories.Builder subfactoriesBuilder
+    ) throws IOException {
+        if (parent != null) {
+            throw new IllegalArgumentException("[random_sampler] aggregation [" + getName() + "] cannot have a parent aggregation");
+        }
+        if (subfactoriesBuilder.getAggregatorFactories().isEmpty()) {
+            throw new IllegalArgumentException("[random_sampler] aggregation [" + getName() + "] must have sub-aggregations");
+        }
+        List<AggregationBuilder> flattenBuilders = new ArrayList<>();
+        recursivelyFlattenAggs(subfactoriesBuilder.getAggregatorFactories(), flattenBuilders);
+        for (AggregationBuilder builder : flattenBuilders) {
+            if (builder instanceof CardinalityAggregationBuilder || builder instanceof NestedAggregationBuilder) {
+                throw new IllegalArgumentException(
+                    "[random_sampler] aggregation ["
+                        + getName()
+                        + "] does not support sampling ["
+                        + builder.getType()
+                        + "] aggregation ["
+                        + builder.getName()
+                        + "]"
+                );
+            }
+        }
+        return new RandomSamplerAggregatorFactory(name, seed, p, context, parent, subfactoriesBuilder, metadata);
+    }
+
+    @Override
+    protected XContentBuilder internalXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field(PROBABILITY.getPreferredName(), p);
+        if (setSeed) {
+            builder.field(SEED.getPreferredName(), seed);
+        }
+        builder.endObject();
+        return null;
+    }
+
+    @Override
+    protected AggregationBuilder shallowCopy(AggregatorFactories.Builder factoriesBuilder, Map<String, Object> metadata) {
+        return new RandomSamplerAggregationBuilder(this, factoriesBuilder, metadata);
+    }
+
+    @Override
+    public BucketCardinality bucketCardinality() {
+        return BucketCardinality.ONE;
+    }
+
+    @Override
+    public String getType() {
+        return NAME;
+    }
+
+    @Override
+    public int hashCode() {
+        if (setSeed) {
+            return Objects.hash(super.hashCode(), p, seed);
+        } else {
+            return Objects.hash(super.hashCode(), p);
+        }
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (obj == null || getClass() != obj.getClass()) return false;
+        if (super.equals(obj) == false) return false;
+        RandomSamplerAggregationBuilder other = (RandomSamplerAggregationBuilder) obj;
+        return Objects.equals(p, other.p) && (setSeed == false || Objects.equals(seed, other.seed));
```

Comment:
You are correct here. Since having the same seed & probability (regardless if the seed was random or set by a user) results in the same data being returned in the aggregator. Consequently, these objects should be considered equal.